Mathlib/CategoryTheory/Limits/Sifted.lean

```diff
@@ -114,4 +126,154 @@ instance [F.Final] [G.Final] : (F.prod' G).Final :=
+
+instance colim_preservesLimitsOfShape_pempty_of_isSifted :
+    PreservesLimitsOfShape (Discrete PEmpty.{1}) (colim : (C ⥤ _) ⥤ Type u) :=
+  preservesLimitsOfShape_pempty_of_preservesTerminal _
+
+/-- If `C` is sifted, the `colim` functor `(C ⥤ Type) ⥤ Type` preserves finite products. -/
+instance colim_preservesFiniteProducts_of_isSifted :
+    PreservesFiniteProducts (colim : (C ⥤ _) ⥤ Type u ) :=
+  PreservesFiniteProducts.of_preserves_binary_and_terminal colim
+
+end
+
+section
+
+open Opposite in
+open scoped MonoidalCategory.ExternalProduct in
+/-- If the `colim` functor `(C ⥤ Type) ⥤ Type` preserves binary products, then `C` is sifted or
+empty. -/
+theorem isSiftedOrEmpty_of_colim_preservesBinaryProducts
+    [PreservesLimitsOfShape (Discrete WalkingPair) (colim : (C ⥤ Type u) ⥤ Type u)] :
+    IsSiftedOrEmpty C := by
+  apply final_of_colimit_comp_coyoneda_iso_pUnit
+  rintro ⟨c₁, c₂⟩
+  calc colimit <| diag C ⋙ coyoneda.obj (op (c₁, c₂))
+    _ ≅ colimit <| _ ⋙ (coyoneda.obj _) ⊠ (coyoneda.obj _) :=
+      HasColimit.isoOfNatIso <| isoWhiskerLeft _ <| .refl _
+    _ ≅ colimit (_ ⊗ _) := HasColimit.isoOfNatIso <| .refl _
+    _ ≅ (colimit _) ⊗ (colimit _) := CartesianMonoidalCategory.prodComparisonIso colim _ _
+    _ ≅ PUnit ⊗ PUnit := (Coyoneda.colimitCoyonedaIso _) ⊗ᵢ (Coyoneda.colimitCoyonedaIso _)
+    _ ≅ PUnit := λ_ _
+
+lemma isSiftedOrEmpty_of_colim_preservesFiniteProducts
+    [h : PreservesFiniteProducts (colim : (C ⥤ Type u) ⥤ Type u)] :
+    IsSiftedOrEmpty C := by
+  rcases Finite.exists_equiv_fin WalkingPair with ⟨_, ⟨e⟩⟩
+  haveI : PreservesLimitsOfShape (Discrete WalkingPair) (colim : (C ⥤ _) ⥤ Type u) :=
+    preservesLimitsOfShape_of_equiv (Discrete.equivalence e.symm) _
+  exact isSiftedOrEmpty_of_colim_preservesBinaryProducts
+
+lemma nonempty_of_colim_preservesLimitsOfShapeFinZero
+    [PreservesLimitsOfShape (Discrete (Fin 0)) (colim : (C ⥤ Type u) ⥤ Type u)] :
+    Nonempty C := by
+  suffices connected : IsConnected C by infer_instance
+  rw [Types.isConnected_iff_colimit_constPUnitFunctor_iso_pUnit]
+  constructor
+  haveI : PreservesLimitsOfShape (Discrete PEmpty) (colim : (C ⥤ _) ⥤ Type u) :=
+    preservesLimitsOfShape_of_equiv (Discrete.equivalence finZeroEquiv') _
+  apply HasColimit.isoOfNatIso (_: Types.constPUnitFunctor C ≅ (⊤_ (C ⥤ Type u)))|>.trans
+  · apply PreservesTerminal.iso colim |>.trans
+    exact Types.terminalIso
+  · apply_rules [IsTerminal.uniqueUpToIso _ terminalIsTerminal, evaluationJointlyReflectsLimits]
+    intro _
+    exact isLimitChangeEmptyCone _ Types.isTerminalPunit _ <| Iso.refl _
+
+/-- If the `colim` functor `(C ⥤ Type) ⥤ Type` preserves finite products, then `C` is sifted. -/
+theorem of_colim_preservesFiniteProducts
+    [h : PreservesFiniteProducts (colim : (C ⥤ Type u) ⥤ Type u)] :
+    IsSifted C := by
+  have := @isSiftedOrEmpty_of_colim_preservesFiniteProducts _ _ h
+  have := @nonempty_of_colim_preservesLimitsOfShapeFinZero _ _ (h.preserves 0)
+  constructor
```

Comment:
It would be a good idea to make `C` an explicit parameter:
```suggestion
    IsSifted C := by
  have := nonempty_of_colim_preservesLimitsOfShapeFinZero (C := C)
  have := isSiftedOrEmpty_of_colim_preservesFiniteProducts (C := C)
```